From 8a64faa325973c322c619a21ea3308016bb43f61 Mon Sep 17 00:00:00 2001
From: Naveen Mamindlapalli <naveenm@marvell.com>
Date: Sat, 15 May 2021 12:50:24 +0530
Subject: [PATCH 1534/1921] octeontx2-af: fix error return code in
 rvu_mbox_handler_set_vf_perm

Fix to return success from rvu_mbox_handler_set_vf_perm() when the
NIXLF is not attached. This happens when PF driver disables VF trust
during reboot where VF driver is unloaded first & when VF is bind
to vfio-pci driver and SR-IOV is disabled by PF.

Change-Id: I26afb3b036c775fdbbfaed9d7b91a44aa5d14d6d
Fixes: efeb72a0cd70 ("octeontx2-af: Add multicast/promisc packet replication table")
Signed-off-by: Naveen Mamindlapalli <naveenm@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/51951
Reviewed-by: Sunil Kovvuri Goutham <sgoutham@marvell.com>
Tested-by: Sunil Kovvuri Goutham <sgoutham@marvell.com>
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 drivers/net/ethernet/marvell/octeontx2/af/rvu.c | 17 ++++++++---------
 1 file changed, 8 insertions(+), 9 deletions(-)

diff --git a/drivers/net/ethernet/marvell/octeontx2/af/rvu.c b/drivers/net/ethernet/marvell/octeontx2/af/rvu.c
index 40c4c0e50b41..e67e0568bc7e 100644
--- a/drivers/net/ethernet/marvell/octeontx2/af/rvu.c
+++ b/drivers/net/ethernet/marvell/octeontx2/af/rvu.c
@@ -2100,21 +2100,20 @@ int rvu_mbox_handler_set_vf_perm(struct rvu *rvu, struct set_vf_perm *req,
 	target = (pcifunc & ~RVU_PFVF_FUNC_MASK) | (req->vf + 1);
 	pfvf = rvu_get_pfvf(rvu, target);
 
-	blkaddr = rvu_get_blkaddr(rvu, BLKTYPE_NIX, target);
-	if (blkaddr < 0)
-		return NPA_AF_ERR_AF_LF_INVALID;
-
-	nixlf = rvu_get_lf(rvu, &hw->block[blkaddr], pcifunc, 0);
-	if (nixlf < 0)
-		return NPA_AF_ERR_AF_LF_INVALID;
-
 	if (req->flags & RESET_VF_PERM)
 		pfvf->flags &= RVU_CLEAR_VF_PERM;
 	else if (test_bit(PF_SET_VF_TRUSTED, &pfvf->flags) ^
 		 (req->flags & VF_TRUSTED)) {
 		change_bit(PF_SET_VF_TRUSTED, &pfvf->flags);
+		/* disable multicast and promisc entries */
 		if (!test_bit(PF_SET_VF_TRUSTED, &pfvf->flags)) {
-			/* Delete multicast and promisc MCAM entries */
+			blkaddr = rvu_get_blkaddr(rvu, BLKTYPE_NIX, target);
+			if (blkaddr < 0)
+				return 0;
+			nixlf = rvu_get_lf(rvu, &hw->block[blkaddr],
+					   target, 0);
+			if (nixlf < 0)
+				return 0;
 			npc_enadis_default_mce_entry(rvu, target, nixlf,
 						     NIXLF_ALLMULTI_ENTRY,
 						     false);
-- 
2.31.1

